sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DemoteBroadcastHashJoin.scala

```diff
@@ -29,11 +28,11 @@ import org.apache.spark.sql.internal.SQLConf
 case class DemoteBroadcastHashJoin(conf: SQLConf) extends Rule[LogicalPlan] {
 
   private def shouldDemote(plan: LogicalPlan): Boolean = plan match {
-    case LogicalQueryStage(_, stage: ShuffleQueryStageExec) if stage.resultOption.isDefined
-      && stage.resultOption.get != null =>
-      val mapOutputStatistics = stage.resultOption.get.asInstanceOf[MapOutputStatistics]
-      val partitionCnt = mapOutputStatistics.bytesByPartitionId.length
-      val nonZeroCnt = mapOutputStatistics.bytesByPartitionId.count(_ > 0)
+    case LogicalQueryStage(_, s: ShuffleQueryStageExec) if s.resultOption.isEmpty => false
+    case LogicalQueryStage(_, s: ShuffleQueryStageExec) if s.mapStats.isDefined =>
```

Comment:
nit: we can still use one case `case LogicalQueryStage(_, s: ShuffleQueryStageExec) if s.resultOption.isDefined && s.mapStates.isDefined`